File: iq/twopointer/palindrome/solution_test.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package palindrome_test

import (
	"testing"

	"github.com/skhal/lab/iq/twopointer/palindrome"
)

func TestIs(t *testing.T) {
	tests := []struct {
		name string
		str  string
		want bool
	}{
		{
			name: "empty",
			want: true,
		},
		{
			name: "one letter",
			str:  "a",
			want: true,
		},
		{
			name: "one digit",
			str:  "1",
			want: true,
		},
		{
			name: "non-alphanumic letter",
			str:  ".",
			want: true,
		},
		{
			name: "two letters",
			str:  "aa",
			want: true,
		},
		{
			name: "two digits",
			str:  "11",
			want: true,
		},
		{
			name: "two distinct letters",
			str:  "ab",
		},
		{
			name: "letter and digit",
			str:  "a1",
		},
		{
			name: "letter and non-alphanumeric",
			str:  "a.",
			want: true,
		},
		{
			name: "digit and non-alphanumeric",
			str:  "1.",
			want: true,
		},
		{
			name: "non-alphanumeric and letter",
			str:  ".a",
			want: true,
		},
		{
			name: "non-alphanumeric and digit",
			str:  ".1",
			want: true,
		},
	}
	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got := palindrome.Is(tc.str)
			if got != tc.want {
				t.Errorf("palindrome.Is(%q) = %v, want %v", tc.str, got, tc.want)
			}
		})
	}
}

File: iq/twopointer/lexiseq/solution_test.cc
// Copyright 2025 Samvel Khalatyan. All rights reserved.
//
//go:build nobuild

#include <cctype>
#include <iomanip>
#include <ostream>
#include <vector>

#include "gtest/gtest.h"
#include "iq/twopointer/lexiseq/solution.h"

namespace iq::twopointer::lexiseq {
namespace {

using ::testing::TestParamInfo;
using ::testing::TestWithParam;
using ::testing::ValuesIn;

constexpr char kCharUnderscore = '_';

struct NextTestParam {
  std::string name;
  std::string s;
  std::string want;

  friend std::ostream& operator<<(std::ostream& os, const NextTestParam& tp) {
    return os << std::quoted(tp.name);
  }
};

class NextTest : public TestWithParam<NextTestParam> {};

TEST_P(NextTest, Test) {
  const NextTestParam& tp = GetParam();

  const std::string got = Next(tp.s);

  EXPECT_EQ(got, tp.want);
}

const std::vector<NextTestParam> kTestParams{
    {.name = "empty", .s = "", .want = ""},
    {.name = "one letter", .s = "a", .want = "a"},
    {.name = "two letters same", .s = "aa", .want = "aa"},
    {.name = "two letters ascending", .s = "ab", .want = "ba"},
    {.name = "two letters descending", .s = "ba", .want = "ab"},
    {.name = "three letters same", .s = "aaa", .want = "aaa"},
    {.name = "three letters first two same", .s = "aab", .want = "aba"},
    {.name = "three letters first and last same", .s = "aba", .want = "baa"},
    {.name = "three letters second and last same", .s = "baa", .want = "aab"},
};

INSTANTIATE_TEST_SUITE_P(NextTest, NextTest, ValuesIn(kTestParams),
                         [](const TestParamInfo<NextTestParam>& info) {
                           std::string name = info.param.name;
                           std::replace_if(
                               name.begin(), name.end(),
                               [](char c) { return !std::isalnum(c); },
                               kCharUnderscore);
                           return name;
                         });

}  // namespace
}  // namespace iq::twopointer::lexiseq
